File: dojo_plugin/product/telemetry.py
import hashlib


FORBIDDEN_FIELDS = frozenset(
    {
        "answer",
        "password",
        "token",
        "flag",
        "submission",
        "submissionbody",
        "chat",
        "chattext",
        "prompt",
        "fullprompt",
        "studenttext",
        "ip",
    }
)

EVENT_FIELDS = frozenset(
    {
        "name",
        "eventVersion",
        "journeyId",
        "mode",
        "surface",
        "resource",
        "action",
        "fromState",
        "toState",
        "result",
        "reasonCode",
        "durationMs",
        "requestId",
        "client",
    }
)

CLIENT_FIELDS = frozenset(
    {"viewportClass", "offline", "connectionType", "reducedMotion"}
)

PRODUCT_EVENT_NAMES = frozenset(
    {
        "resource_action_completed",
        "unknown_state",
        "recovery_selected",
        "view_state_changed",
        "state_contradiction",
        "request_failed",
        "route_redirected",
        "performance_vital",
    }
)


def _normalized_key(value):
    return str(value).replace("_", "").replace("-", "").casefold()


def _all_keys(value):
    if isinstance(value, dict):
        for key, child in value.items():
            yield _normalized_key(key)
            yield from _all_keys(child)
    elif isinstance(value, (list, tuple)):
        for child in value:
            yield from _all_keys(child)


def opaque_scope_id(value, salt="aisecedu-product-telemetry"):
    return hashlib.sha256(f"{salt}:{value}".encode()).hexdigest()[:20]


def sanitize_event(payload):
    source = payload if isinstance(payload, dict) else {}
    lower_keys = set(_all_keys(source))
    if lower_keys & FORBIDDEN_FIELDS:
        raise ValueError("事件包含不允许采集的内容。")
    result = {key: source[key] for key in EVENT_FIELDS if key in source}
    name = str(result.get("name") or "").strip().lower()
    if name not in PRODUCT_EVENT_NAMES:
        raise ValueError("事件名称不在隐私白名单中。")
    result["name"] = name
    result["eventVersion"] = max(1, min(10, int(result.get("eventVersion") or 1)))
    resource = result.get("resource")
    if isinstance(resource, dict):
        resource_id = resource.get("id")
        id_hash = resource.get("idHash")
        result["resource"] = {
            "type": str(resource.get("type") or "resource")[:48],
            "idHash": (
                opaque_scope_id(resource_id)
                if resource_id is not None
                else str(id_hash or "")[:64] or None
            ),
        }
    elif "resource" in result:
        result.pop("resource")
    client = result.get("client")
    if isinstance(client, dict):
        result["client"] = {
            key: client[key]
            for key in CLIENT_FIELDS
            if key in client
        }
        for key in ("offline", "reducedMotion"):
            if key in result["client"]:
                result["client"][key] = bool(result["client"][key])
        for key in ("viewportClass", "connectionType"):
            if key in result["client"]:
                result["client"][key] = str(result["client"][key])[:32]
    elif "client" in result:
        result.pop("client")
    for key in ("journeyId", "mode", "surface", "action", "fromState", "toState", "result", "reasonCode", "requestId"):
        if key in result and result[key] is not None:
            result[key] = str(result[key])[:160]
    if "durationMs" in result:
        result["durationMs"] = max(0, min(86_400_000, int(result["durationMs"] or 0)))
    return result

File: dojo_theme/static/js/dojo/core/telemetry.js
(function () {
  "use strict";

  const root = window.AISecEdu = window.AISecEdu || {};
  const names = new Set([
    "resource_action_completed", "unknown_state", "recovery_selected",
    "view_state_changed", "state_contradiction", "request_failed",
    "route_redirected", "performance_vital",
  ]);
  const fields = new Set([
    "name", "eventVersion", "journeyId", "mode", "surface", "resource",
    "action", "fromState", "toState", "result", "reasonCode", "durationMs",
    "requestId", "client",
  ]);
  const forbidden = new Set([
    "answer", "password", "token", "flag", "submission", "submissionbody",
    "chat", "chattext", "prompt", "fullprompt", "studenttext", "ip",
  ]);
  const seen = new Map();
  let sent = 0;

  function normalizedKey(value) {
    return String(value || "").replace(/[_-]/g, "").toLowerCase();
  }

  function containsForbidden(value) {
    if (Array.isArray(value)) return value.some(containsForbidden);
    if (!value || typeof value !== "object") return false;
    return Object.entries(value).some(([key, child]) => (
      forbidden.has(normalizedKey(key)) || containsForbidden(child)
    ));
  }

  function journeyId() {
    const key = "aisecedu-telemetry-journey";
    try {
      let value = sessionStorage.getItem(key);
      if (!value) {
        value = window.crypto?.randomUUID?.()
          || `${Date.now()}-${Math.random().toString(16).slice(2)}`;
        sessionStorage.setItem(key, value);
      }
      return value;
    } catch (error) {
      return "ephemeral";
    }
  }

  function clientContext() {
    const width = window.innerWidth;
    return {
      viewportClass: width < 480 ? "phone" : width < 900 ? "tablet" : "desktop",
      offline: navigator.onLine === false,
      connectionType: navigator.connection?.effectiveType || "unknown",
      reducedMotion: window.matchMedia?.("(prefers-reduced-motion: reduce)")?.matches || false,
    };
  }

  function sanitize(payload) {
    if (!payload || typeof payload !== "object" || containsForbidden(payload)) return null;
    const name = String(payload.name || "").trim().toLowerCase();
    if (!names.has(name)) return null;
    const event = {};
    Object.entries(payload).forEach(([key, value]) => {
      if (fields.has(key)) event[key] = value;
    });
    event.name = name;
    event.eventVersion = 1;
    event.journeyId = String(event.journeyId || journeyId()).slice(0, 160);
    event.mode = String(event.mode || document.body?.dataset.productMode || "public").slice(0, 32);
    event.surface = String(event.surface || document.body?.dataset.telemetrySurface || event.mode).slice(0, 80);
    event.client = clientContext();
    if (event.resource && typeof event.resource === "object") {
      event.resource = {
        type: String(event.resource.type || "resource").slice(0, 48),
        id: String(event.resource.id || "").slice(0, 160) || undefined,
      };
    } else {
      delete event.resource;
    }
    return event;
  }

  function track(name, details) {
    if (sent >= 40 || !root.request) return Promise.resolve(false);
    const event = sanitize({name, ...(details || {})});
    if (!event) return Promise.resolve(false);
    const signature = JSON.stringify([
      event.name, event.surface, event.action, event.fromState,
      event.toState, event.result, event.reasonCode,
    ]);
    const last = seen.get(signature) || 0;
    if (Date.now() - last < 1200) return Promise.resolve(false);
    seen.set(signature, Date.now());
    sent += 1;
    return root.request("/ui/telemetry", {
      method: "POST",
      json: event,
      timeoutMs: 5000,
      retries: 0,
      dedupeKey: false,
      invalidateCache: false,
    }).then(() => true).catch(() => false);
  }

  document.addEventListener("aisecedu:unknown-state", event => {
    track("unknown_state", {
      action: String(event.detail?.domain || "state"),
      toState: String(event.detail?.state || "unknown"),
      result: "safe_fallback",
    });
  });
  document.addEventListener("aisecedu:recovery", event => {
    track("recovery_selected", {
      action: "recovery",
      reasonCode: event.detail?.code,
      requestId: event.detail?.requestId,
    });
  });
  document.addEventListener("aisecedu:async-state", event => {
    track("view_state_changed", {toState: event.detail?.state});
  });
  ["batch", "workspace"].forEach(kind => {
    document.addEventListener(`aisecedu:${kind}-contradiction`, () => {
      track("state_contradiction", {action: kind, result: "blocked_success_claim"});
    });
  });

  root.telemetry = {track, sanitize, names: Array.from(names)};
  const queued = Array.isArray(root.telemetryQueue) ? root.telemetryQueue.splice(0) : [];
  queued.slice(0, 40).forEach(item => track(item?.name, item?.details));
})();
